File: src/main/java/com/soapdogg/lispInterpreter/interpreter/Interpreter.java
package com.soapdogg.lispInterpreter.interpreter;

import com.soapdogg.lispInterpreter.datamodels.Node;
import com.soapdogg.lispInterpreter.datamodels.PartitionedRootNodes;
import com.soapdogg.lispInterpreter.datamodels.Token;
import com.soapdogg.lispInterpreter.datamodels.UserDefinedFunction;
import com.soapdogg.lispInterpreter.evaluator.ProgramEvaluator;
import com.soapdogg.lispInterpreter.generator.UserDefinedFunctionGenerator;
import com.soapdogg.lispInterpreter.parser.RootParser;
import com.soapdogg.lispInterpreter.printer.ListNotationPrinter;
import com.soapdogg.lispInterpreter.tokenizer.Tokenizer;
import lombok.AllArgsConstructor;

import java.util.HashMap;
import java.util.List;
import java.util.Scanner;

@AllArgsConstructor(staticName = "newInstance")
public final class Interpreter{

	private final Tokenizer tokenizer;
	private final RootParser rootParser;
	private final ProgramEvaluator program;
	private final RootNodePartitioner rootNodePartitioner;
	private final UserDefinedFunctionGenerator userDefinedFunctionGenerator;
	private final ListNotationPrinter listNotationPrinter;

	public String interpret(Scanner in) throws Exception{
	    List<Token> tokens = tokenizer.tokenize(in);
	    List<Node> rootNodes = rootParser.parse(tokens);
		PartitionedRootNodes partitionedRootNodes = rootNodePartitioner.partitionRootNodes(
			rootNodes
		);
		List<UserDefinedFunction> userDefinedFunctions = userDefinedFunctionGenerator.generateUserDefinedFunctions(
			partitionedRootNodes.getDefunNodes()
		);
		List<Node> evaluatedNodes = program.evaluate(
			partitionedRootNodes.getEvaluatableNodes(),
			userDefinedFunctions,
			new HashMap<>()
		);
		return listNotationPrinter.printInListNotation(evaluatedNodes);
    }
}

 

File: src/main/java/com/soapdogg/lispInterpreter/interpreter/Interpreter.kt
package com.soapdogg.lispInterpreter.interpreter

import com.soapdogg.lispInterpreter.asserter.ExpressionListLengthAsserter
import com.soapdogg.lispInterpreter.evaluator.ProgramEvaluator
import com.soapdogg.lispInterpreter.generator.UserDefinedFunctionGenerator
import com.soapdogg.lispInterpreter.parser.RootParser
import com.soapdogg.lispInterpreter.printer.ListNotationPrinter
import com.soapdogg.lispInterpreter.tokenizer.Tokenizer
import java.util.Scanner

class Interpreter(
    private val tokenizer: Tokenizer,
    private val rootParser: RootParser,
    private val rootNodePartitioner: RootNodePartitioner,
    private val userDefinedFunctionGenerator: UserDefinedFunctionGenerator,
    private val expressionListLengthAsserter: ExpressionListLengthAsserter,
    private val programEvaluator: ProgramEvaluator,
    private val listNotationPrinter: ListNotationPrinter
){

    fun interpret(
        scanner: Scanner
    ): String {
        val tokens = tokenizer.tokenize(scanner)
        val rootNodes = rootParser.parse(tokens)

        val partitionedRootNodes = rootNodePartitioner.partitionRootNodes(
            rootNodes
        )
        val userDefinedFunctions = partitionedRootNodes.defunNodes.map{
            userDefinedFunctionGenerator.evaluateLispFunction(it) }.toMap()

        expressionListLengthAsserter.assertLengthIsAsExpected(
            partitionedRootNodes.evaluatableNodes,
            userDefinedFunctions
        )

        val evaluatedNodes = programEvaluator.evaluatePostOrder(
            partitionedRootNodes.evaluatableNodes,
            userDefinedFunctions
        )

        return listNotationPrinter.printInListNotation(evaluatedNodes)
    }
}